### PR TITLE
Fix #169: `pre` and `post` hooks to canvas renderers

### LINKED ISSUE #169: Add possibility to have `pre` and `post` hooks to canvas renderers
Add the possibility for the `def` method in the canvas renderers to have `pre` and `post` hooks:

```
xyz.def = {
   pre: function(...){ this.font = "";},
   render: function(..){..use this.font as a context.font cache...},
   post: function(...){..clean things..},
}
```

The previous renderers.def won't need to be changed, this is gonna be compatible with the two models

It's to solve #167 and #168, and also because it would be nice for future optimizations (batch rendering for example).

Other possible solution:

* Pass all the objects to render together in one `renderAll` function to the renderer: More generic but would be harder to implement and wouldn't integrate well with the previous model

### PR DESCRIPTION
**Pull request by mdamien**

### CODE REVIEW

**Comment by sheymann** [comment]
Can you edit to merge into linkurious-version instead of master please?

**Comment by sheymann** [comment]
- [ ] please run `grunt` and comply with the linter rules (e.g. max 80 chars per line)

File: src/renderers/sigma.renderers.canvas.js [hunk 1]
```diff
@@ -103,7 +103,81 @@
...
+   * 
+   * @param  {array}                 renderers  Renderers to use
+   * @param  {string}                type       "edges" or "nodes"
+   * @param  {array}                 elements   Elements to render
+   * @param  {Context2D}             ctx        Ctontext to draz on
+   * @param  {integer}               start      Starting index of the elements to render
+   * @param  {integer}               end        Last index of the elements to render
+   * @param  {object}                options    An object of options.
+   */
+  sigma.renderers.canvas.prototype.renderRenderers = function(renderers, type, elements, ctx, start, end, options) {
```
Comment (sheymann) [inline src/renderers/sigma.renderers.canvas.js]: Rename to `applyRenderers`?

File: src/renderers/sigma.renderers.canvas.js [hunk 2]
```diff
@@ -103,7 +103,81 @@
   };
 
 
+  /**
+   * Render elements with the given renderers
+   * 
+   * @param  {array}                 renderers  Renderers to use
+   * @param  {string}                type       "edges" or "nodes"
+   * @param  {array}                 elements   Elements to render
```
Comment (sheymann) [inline src/renderers/sigma.renderers.canvas.js]: What does "elements" refer to?

File: src/renderers/sigma.renderers.canvas.js [hunk 3]
```diff
@@ -103,7 +103,81 @@
...
+   * @param  {Context2D}             ctx        Ctontext to draz on
```
Comment (sheymann) [inline src/renderers/sigma.renderers.canvas.js]: Typo "Ctontext"

File: src/renderers/sigma.renderers.canvas.js (hunk 2 above)
Comment (mdamien) [inline src/renderers/sigma.renderers.canvas.js]: It was edges or nodes but I'm changing it to `edges/nodes`

File: src/renderers/sigma.renderers.canvas.js [hunk 4]
```diff
@@ -103,7 +103,81 @@
...
+   * @param  {integer}               start      Starting index of the elements to render
```
Comment (sheymann) [inline src/renderers/sigma.renderers.canvas.js]: What are indexes?

File: src/renderers/sigma.renderers.canvas.js [hunk 5]
```diff
@@ -103,7 +103,81 @@
...
+   * @param  {integer}               end        Last index of the elements to render
+   * @param  {object}                options    An object of options.
```
Comment (sheymann) [inline src/renderers/sigma.renderers.canvas.js]: Which options are available?

**Comment by mdamien** [comment]
Another thing: Hover renderers have their own rendering routine outside of the renderers

https://github.com/Linkurious/linkurious.js/blob/linkurious-version/src/misc/sigma.misc.drawHovers.js#L53-L218

**Pull request closed without merging**